apis/r/R/SOMADataFrame.R

```diff
@@ -339,6 +339,30 @@ SOMADataFrame <- R6::R6Class(
       class = 'notYetImplementedError'
     )),
 
+    #' @description Returns a named list of minimum/maximum pairs, one per index
+    #' column, currently storable on each index column of the dataframe. These
+    #' can be resized up to `maxdomain`.
+    #' (lifecycle: maturing)
+    #' @return Named list of minimum/maximum values.
+    domain = function() {
+      as.list(
+        arrow::as_record_batch(
+          arrow::as_arrow_table(
+            domain(self$uri, private$.soma_context))))
```

Comment:
What's the R-object return type of `domain()` (and `maxdomain()` and `non_empty_domain()`)? Why do we take the results and go to an Arrow Table, then an Arrow RecordBatch, then a list? Do we need these conversions in the first place? I know that you can call `as.list()` on an Arrow Table directly, but do we need to go through a Table in the first place?

```R
> (tbl <- arrow::as_arrow_table(mtcars))
Table
32 rows x 11 columns
$mpg <double>
$cyl <double>
$disp <double>
$hp <double>
$drat <double>
$wt <double>
$qsec <double>
$vs <double>
$am <double>
$gear <double>
$carb <double>

See $metadata for additional Schema metadata
> identical(as.list(tbl), as.list(arrow::as_record_batch(tbl)))
[1] TRUE
```

Reply:
I found that these steps were all necessary. Errors otherwise. I can share details if you like.

Reply:
The R object returned is a named list of two-element vectors.

Reply:
Please do share details; it doesn't make sense for a list-return to throw errors or for `as.list(tbl)` to fail while `as.list(arrow::as_record_batch(tbl))` works

Reply:
@mojaveazure 

```
# This is a local copy of pbmc3k
> sdf <- SOMAOpen('/var/p/obs')

> sdf$schema()
Schema
soma_joinid: int64 not null
obs_id: string not null
n_genes: int64 not null
percent_mito: float not null
n_counts: float not null
louvain: dictionary<values=string, indices=int8> not null

> ctx <- tiledbsoma:::soma_context()

> a <- domain(sdf$uri, ctx)

> a
<nanoarrow_array struct[0]>
 $ length    : int 0
 $ null_count: int 0
 $ offset    : int 0
 $ buffers   :List of 1
  ..$ :<nanoarrow_buffer validity<bool>[0][0 b]> ``
 $ children  :List of 1
  ..$ soma_joinid:<nanoarrow_array int64[2]>
  .. ..$ length    : int 2
  .. ..$ null_count: int 0
  .. ..$ offset    : int 0
  .. ..$ buffers   :List of 2
  .. .. ..$ :<nanoarrow_buffer validity<bool>[0][0 b]> ``
  .. .. ..$ :<nanoarrow_buffer data<int64>[2][16 b]> `0.00e+00 2.15e+09`
  .. ..$ dictionary: NULL
  .. ..$ children  : list()
 $ dictionary: NULL

> as.list(a)
Error in as.vector.nanoarrow_array(x, "list") :
  identical(mode, "any") is not TRUE

> as.list(arrow::as_record_batch(a))
Error in `[<-`:
! Assigned data `map(.subset(x, unname), vectbl_set_names, NULL)` must be compatible with existing data.
✖ Existing data has 0 rows.
✖ Assigned data has 2 rows.
ℹ Only vectors of size 1 are recycled.
Caused by error in `vectbl_recycle_rhs_rows()`:
! Can't recycle input of size 2 to size 0.
Run `rlang::last_trace()` to see where the error occurred.

> as.list(arrow::as_arrow_table(a))
Error in `[<-`:
! Assigned data `map(.subset(x, unname), vectbl_set_names, NULL)` must be compatible with existing data.
✖ Existing data has 0 rows.
✖ Assigned data has 2 rows.
ℹ Only vectors of size 1 are recycled.
Caused by error in `vectbl_recycle_rhs_rows()`:
! Can't recycle input of size 2 to size 0.
Run `rlang::last_trace()` to see where the error occurred.

> as.list(arrow::as_record_batch(arrow::as_arrow_table(a)))
$soma_joinid
[1]          0 2147483646
```

Reply:
Ah, okay. It seems like `domain()` et al don't actually read in the domain but leave the domain information in a nanoarrow buffer. I don't know if going from nanoarrow array w/ buffer -> Arrow Table -> Arrow RecordBatch -> list is the best way forward, but I can't find a way around it in R (especially since it seems like the RecordBatch is needed to get the data out of the buffer)

Reply:
My question now is do we want these returns to be a list or a vector? Another way of putting that is will these functions always return two values (min, max) or potentially multiple sets of min/max pairs?

Reply:
@mojaveazure that's partly a UX question

* Each slot's _value_ will always be a 2-element lo/hi pair -- one per dim -- no changing that
* What's a UX questiion is the _keys_ -- what's better, a named list (which I personally like), or, an unnamed list, or, an outer vector of length N where N is the number of dims

I personally like the above three choices in the order I listed them, b/c (a) there's more info there; (b) people can remove info by `unname(dom)` or `as.vector(x)` but it's harder to put the info back in if we just give them a length-N vector of length-2 vectors

But that's just my general feel as a polyglottal language user; I defer to you

I'd also point out that on the Python side it's a length-n tuple of length-2 tuples ... so for strict partity we might go that way ...

Reply:
Maybe have an optional `as.something` arg to the domainish accessors where the default is the compact way but people can get the named-list info if they ask for it? Just a thought. (And it doesn't add much complexity.)

Reply:
Vectors are strictly flat and cannot be nested. Because it seems like the base unit of return is the dim and there's effectively one return (a min/max pair) per dim, a list is the appropriate return. I agree that named is better than unnamed as it's easer to unname than to re-name

Reply:
(I apologize for the back-and-forth on this; I was not part of the discussions for these new methods, so I'm learning them for the first time here)